DiverseSelector/selectors.py

```diff
@@ -337,15 +419,86 @@ def select_from_cluster(self, arr, num_selected, indices=None):
             raise ValueError(f"{self.grid_method} not a valid grid_method")
 
         old_len = 0
+        to_delete = []
         rng = np.random.default_rng(seed=self.random_seed)
         while len(selected) < num_selected:
-            for bin_idx, bin_list in bins.items():
-                if len(bin_list) > 0:
+            num_needed = num_selected - len(selected)
+            bin_count = len(bins)
+            if bin_count <= num_needed:
+                for bin_idx, bin_list in bins.items():
                     random_int = rng.integers(low=0, high=len(bin_list), size=1)[0]
                     mol_id = bin_list.pop(random_int)
                     selected.append(mol_id)
+                    if len(bin_list) == 0:
+                        to_delete.append(bin_idx)
+                for idx in to_delete:
+                    del bins[idx]
+                to_delete = []
 
+            else:
+                diversity = []
+                for bin_idx, bin_list in bins.items():
+                    diversity.append((compute_diversity(arr[bin_list]), bin_idx))
+                diversity.sort(reverse=True)
+                for _, bin_idx in diversity[:num_needed]:
+                    random_int = rng.integers(low=0, high=len(bins[bin_idx]), size=1)[0]
+                    mol_id = bins[bin_idx].pop(random_int)
+                    selected.append(mol_id)
             if len(selected) == old_len:
                 break
             old_len = len(selected)
         return selected
+
+
+def predict_radius(obj: Union[DirectedSphereExclusion, OptiSim], arr, num_selected, cluster_ids=None):
+    """
+    Algorithm that uses sphere_exclusion for selecting points from cluster.
+    Parameters
+    ----------
+    obj: object
+        instance of dissimilarity selection class
+    arr: np.ndarray
+        coordinate array of points
+    num_selected: int
+        number of molecules that need to be selected.
+    cluster_ids: np.ndarray
+        indices of molecules that form a cluster
+
+    Returns
+    -------
+    selected: list
+        list of ids of selected molecules
+    """
+    if cluster_ids is not None:
+        arr = arr[cluster_ids]
+    if obj.r is not None:
+        return obj.algorithm(arr)
+    # Use numpy.optimize.bisect instead
+    arr_range = (
+        max(arr[:, 0]) - min(arr[:, 0]),
+        max(arr[:, 1]) - min(arr[:, 1]),
+    )
```

Comment:
Maybe this can make the codes faster, https://numpy.org/doc/stable/reference/generated/numpy.ptp.html?

Reply:
Thanks for the suggestion, I've implemented it